lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1580,16 +1611,32 @@ class HSAQueue final : public KalmarQueue
 
             for (int i=0; i<count; i++) {
                 if (barrier->depAsyncOps[i] != nullptr) {
-                    // _needs_sys_release is set when a queue executes a kernel.  It indicates the queue needs to execute a release-to-system
-                    // before host can see the data - this is important for kernels which write non-coherent zero-copy host memory.
-                    // If creating a dependency on a queue which needs_system_release, copy that state here.   If the host then 
-                    // waits on the freshly created marker, runtime will issue a system-release fence.
-                    //
                     auto depHSAQueue = static_cast<Kalmar::HSAQueue *> (barrier->depAsyncOps[i]->getQueue());
-
-                    if (depHSAQueue->_needs_sys_release) {
-                        setNeedsSysRelease(true);
+                    // Same accelerator:
+                    // Inherit system-acquire and system-release bits from the queue which contains the op we are dependent on.
+                    //
+                    // _nextSyncNeedsSysRelease is set when a queue executes a kernel.  
+                    // It indicates the queue needs to execute a release-to-system
+                    // before host can see the data - this is important for kernels which write 
+                    // non-coherent zero-copy host memory.
+                    // If creating a dependency on a queue which needs_system_release, copy that 
+                    // state here.   If the host then waits on the freshly created marker, 
+                    // runtime will issue a system-release fence.
+                    if (depHSAQueue->nextKernelNeedsSysAcquire()) {
+                        setNextKernelNeedsSysAcquire(true);
+                    }
+                    if (depHSAQueue->nextSyncNeedsSysRelease()) {
+                        setNextSyncNeedsSysRelease(true);
                     }
+
+                    if (depHSAQueue->getHSADev() != this->getHSADev()) {
+                        // Cross-accelerator dependency case.
+                        // This requires system-scope acquire 
+                        // TODO - only needed if these are peer GPUs, could optimize with an extra check
+                        DBOUT(DB_WAIT, "  Adding cross-accelerator system-scope acquire\n");
+                        barrier->_acquire_scope = hc::system_scope;
```

Comment:
@bensander this seems to be the only place which writes to `HSABarrier::_acquire_scope` outside `HSABarrier`. Perhaps we should change it to an accessor member function?